cdc/puller/entry_group.go

```diff
@@ -21,43 +21,47 @@ import (
 
 // EntryGroup stores RawKVEntry and corresponding ts
 type EntryGroup struct {
-	entriesMap map[uint64][]*model.RawKVEntry
-	sortedTs   []uint64
+	sortedEntries []*model.RawRowGroup
+	collectTxn    bool
 }
 
 // NewEntryGroup creates a new EntryGroup
-func NewEntryGroup() *EntryGroup {
+func NewEntryGroup(collectTxn bool) *EntryGroup {
 	return &EntryGroup{
-		entriesMap: make(map[uint64][]*model.RawKVEntry),
+		collectTxn: collectTxn,
 	}
 }
 
 // AddEntry adds an RawKVEntry to the EntryGroup, this method *IS NOT* thread safe.
 func (eg *EntryGroup) AddEntry(ts uint64, entry *model.RawKVEntry) {
-	i := sort.Search(len(eg.sortedTs), func(i int) bool { return eg.sortedTs[i] >= ts })
-	if i < len(eg.sortedTs) && eg.sortedTs[i] == ts {
-		eg.entriesMap[ts] = append(eg.entriesMap[ts], entry)
-	} else {
-		eg.sortedTs = append(eg.sortedTs, 0)
-		copy(eg.sortedTs[i+1:], eg.sortedTs[i:])
-		eg.sortedTs[i] = ts
-		eg.entriesMap[ts] = []*model.RawKVEntry{entry}
+	i := sort.Search(len(eg.sortedEntries), func(i int) bool { return eg.sortedEntries[i].Ts >= ts })
+	if i >= len(eg.sortedEntries) || eg.sortedEntries[i].Ts != ts {
+		eg.sortedEntries = append(eg.sortedEntries, nil)
+		copy(eg.sortedEntries[i+1:], eg.sortedEntries[i:])
+		eg.sortedEntries[i] = &model.RawRowGroup{Ts: ts, IsCompleteTxn: true}
 	}
+	eg.sortedEntries[i].Entries = append(eg.sortedEntries[i].Entries, entry)
 }
 
 // Consume retrieves all RawKVEntry with ts no larger than resolvedTs,
 // returns RawTxn list sorted by ts according these RawKVEntry and
 // removes these RawKVEntry from EntryGroup. This method *IS NOT* thread safe.
-func (eg *EntryGroup) Consume(resolvedTs uint64) (txns []model.RawTxn) {
+func (eg *EntryGroup) Consume(resolvedTs uint64) (txns []model.RawRowGroup) {
 	i := 0
-	for ; i < len(eg.sortedTs); i++ {
-		ts := eg.sortedTs[i]
-		if ts > resolvedTs {
+	for ; i < len(eg.sortedEntries); i++ {
+		if eg.sortedEntries[i].Ts > resolvedTs {
 			break
 		}
-		txns = append(txns, model.RawTxn{Ts: ts, Entries: eg.entriesMap[ts]})
-		delete(eg.entriesMap, ts)
+		txns = append(txns, *eg.sortedEntries[i])
 	}
-	eg.sortedTs = eg.sortedTs[i:]
+	if !eg.collectTxn {
```

Comment:
It seems this is never turned off now, maybe we can remove it and add it back with an efficient implementation when we do need it?

Reply:
When the downstream is kafka, we will trun `collectTxn` off .
I will use it in the next pr